backend/cfg/dataflow.ml

```diff
@@ -0,0 +1,218 @@
+      incr iteration;
+      let block = remove_and_return cfg work_set in
+      let predecessors'outputs =
+        ListLabels.concat_map (Cfg.predecessor_labels block)
+          ~f:(fun predecessor_label ->
+            let predecessor_block = Cfg.get_block_exn cfg predecessor_label in
+            let if_among_successors ~normal ~exn ~f =
+              let successor_labels =
+                Cfg.successor_labels predecessor_block ~normal ~exn
+              in
+              if Label.Set.mem block.start successor_labels
+              then [f (Label.Tbl.find res predecessor_label)]
+              else []
+            in
+            if_among_successors ~normal:true ~exn:false ~f:(fun x -> x.after)
+            @ if_among_successors ~normal:false ~exn:true ~f:(fun x ->
+                  x.exception_))
+        |> List.filter_map Fun.id
+      in
+      let old_value =
+        match predecessors'outputs with
+        | [] -> Label.Tbl.find res block.start
+        | hd :: tl ->
+          let before = ListLabels.fold_left tl ~init:hd ~f:D.join in
+          { (Label.Tbl.find res block.start) with before }
+      in
+      let new_value = transfer_block old_value block in
+      Label.Tbl.replace res block.start new_value;
+      let successors_to_visit =
+        Cfg.successor_labels
+          ~normal:(different old_value.after new_value.after)
+          ~exn:(different old_value.exception_ new_value.exception_)
+          block
+      in
+      work_set := Label.Set.union !work_set successors_to_visit
+    done;
+    res
+end
+
+(* CR xclerc for xclerc: move the code below to another module *)
+
+module Domain = struct
+  type t =
+    | Reachable
+    | Unreachable
+
+  let bot = Unreachable
+
+  let equal left right =
+    match left, right with
+    | Reachable, Reachable | Unreachable, Unreachable -> true
+    | Reachable, Unreachable | Unreachable, Reachable -> false
+
+  let lessequal left right =
+    match left, right with
+    | Reachable, Reachable -> true
+    | Reachable, Unreachable -> false
+    | Unreachable, Reachable -> true
+    | Unreachable, Unreachable -> true
+
+  let join left right =
+    match left, right with
+    | Reachable, (Reachable | Unreachable) | Unreachable, Reachable -> Reachable
+    | Unreachable, Unreachable -> Unreachable
+
+  let to_string = function
+    | Reachable -> "reachable"
+    | Unreachable -> "unreachable"
+end
+
+module Transfer = struct
+  type domain = Domain.t
+
+  let basic domain _ = domain
+
+  let terminator domain _ = domain
+
+  let exception_ domain = domain
+end
+
+module Dead_code = Forward (Domain) (Transfer)
+
+let run_dead_code : Cfg_with_layout.t -> unit =
+ fun cfg_with_layout ->
+  let cfg = Cfg_with_layout.cfg cfg_with_layout in
+  let is_entry_label label = Label.equal label cfg.entry_label in
+  let is_trap_handler label = (Cfg.get_block_exn cfg label).is_trap_handler in
+  let init { Cfg.start; _ } =
+    if is_entry_label start || is_trap_handler start
```

Comment:
It is considering the trap handler as reachable to mimic
the current behaviour; it was not there at first, but somehow
broke some CFG invariants.

Reply:
hmm.. is it because of the `Lpushtrap label` that refers to them, or in comparison to the existing dead block elimination that conservatively keeps ahold of all trap handler blocks?